server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -107,8 +108,59 @@ public class BalancedShardsAllocator implements ShardsAllocator {
         "cluster.routing.allocation.move.primary_first",
         false,
         Property.Dynamic,
+        Property.NodeScope,
+        Property.Deprecated
+    );
+
+    /**
+     * ShardMovementStrategy defines the order in which shard movement occurs.
+     * <p>
+     * Allocation settings can have the following values (non-casesensitive):
+     *  <ul>
+     *      <li> <code>NO_PREFERENCE</code> - default behavior in which order of shard movement doesn't matter.
+     *      <li> <code>PRIMARY_FIRST</code> - primary shards are moved first.
+     *      <li> <code>REPLICA_FIRST</code> - replica shards are moved first.
+     *  </ul>
+     * ShardMovementStrategy values or rather their string representation to be used with
+     * {@link BalancedShardsAllocator#SHARD_MOVEMENT_STRATEGY_SETTING} via cluster settings.
+     */
+    public enum ShardMovementStrategy {
```

Comment:
nit - lets pull this to a separate file given its used outside of BalancedShardsAllocator?